File: functions/src/algorithms/pagerank.ts
export const pageRank = (graph: Array<Array<number>>): number[] => {
  let n: number = graph.length;
  let R: number[][] = Array(n);
  for (let i: number = 0; i < n; i++) {
    R[i] = Array(n).fill(0);
  }
  graph.forEach((nodes, i) => {
    if (nodes.length == 0) return;
    let value: number = 1 / nodes.length;
    nodes.forEach(node => (R[node][i] = value));
  });

  console.log(R);

  let probabilityMatrix: number[] = Array(n).fill(1 / n);
  for (let i: number = 0; i < 99; i++) {
    // FUTURE: this is a temporary solution, increase efficiency of matrix multiplication in the future.
    let newProbabilityMatrix: number[] = Array(n);
    R.forEach((scores, i) => {
      let sum: number = 0;
      for (let j: number = 0; j < n; j++) {
        sum += scores[j] * probabilityMatrix[j];
      }
      newProbabilityMatrix[i] = sum;
    });
    probabilityMatrix = newProbabilityMatrix;
  }

  console.log(probabilityMatrix);

  return probabilityMatrix;
};

File: lib/utils/algorithms/page-rank.dart
List<double> pageRank(List<List<int>> graph) {
  final int n = graph.length;
  final List<List<double>> r = List.generate(
    n,
    (_) => List.filled(n, 0, growable: false),
    growable: false,
  );
  graph.asMap().forEach((i, nodes) {
    final double value = 1 / nodes.length;
    nodes.forEach((node) => r[node][i] = value);
  });

  List<double> probabilityMatrix = List.filled(n, 1 / n, growable: false);
  for (int i = 0; i < 99; i++) {
    // FUTURE: this is a temporary solution, increase speed of matrix multiplication in the future.
    final List<double> newProbabilityMatrix = [];
    r.asMap().forEach((i, scores) {
      double sum = 0;
      for (int j = 0; j < n; j++) {
        sum += scores[j] * probabilityMatrix[j];
      }
      newProbabilityMatrix.add(sum); // at i
    });
    probabilityMatrix = newProbabilityMatrix;
  }

  return probabilityMatrix;
}
